File: examples/typescript/good/dip.ts
interface MessageSender {
  send(message: string): void;
}

class GoodTeamsBot implements MessageSender {
  send(message: string): void {
    console.log(`send '${message}' to teams`);
  }
}

class GoodSlackBot implements MessageSender {
  channel: string;
  constructor(channel: string) {
    this.channel = channel;
  }

  send(message: string): void {
    console.log(`send '${message}' to teams`);
  }
}

class GoodAlertService {
  sender: MessageSender;
  constructor(sender: MessageSender) {
    this.sender = sender;
  }

  alert(message: string): void {
    this.sender.send(message);
  }
}

(function () {
  const slackBot = new GoodSlackBot("HAPPY_CHANNEL");
  const alert_service = new GoodAlertService(slackBot);
  alert_service.alert("Hello, World");
})();

File: examples/python/good/dip.py
import abc

class MessageSender(abc.ABC):
  @abc.abstractmethod
  def send(self, message):
      pass

class TeamsBot(MessageSender):
    def send(self, message):
        print(f"send '{message}' to teams")

class SlackBot(MessageSender):
    def __init__(self, channel):
        self.channel = channel

    def send(self, message):
        print(f"send '{message}' to slack")

class AlertService:
  def __init__(self, sender):
    self.sender = sender

  def alert(self, message):
    sender.send(message)

sender = TeamsBot()
alert_service = AlertService(sender)
alert_service.alert("HELLO, WORLD")